From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nrudenko <nikita.rudenko@intel.com>
Date: Wed, 24 Jun 2020 22:23:45 +0300
Subject: [PATCH 4/8] Add pdb info for SPIRVDLL in release mode

---
 lib/SPIRV/CMakeLists.txt | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/lib/SPIRV/CMakeLists.txt b/lib/SPIRV/CMakeLists.txt
index 00000000..00000000 100755
--- a/lib/SPIRV/CMakeLists.txt
+++ b/lib/SPIRV/CMakeLists.txt
@@ -64,6 +64,22 @@ add_llvm_install_targets(install-SPIRVDLL
 
 set_property(GLOBAL APPEND PROPERTY LLVM_EXPORTS SPIRVDLL)
 
+
+if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC" AND CMAKE_BUILD_TYPE MATCHES "Release")
+   target_compile_options(SPIRVDLL PRIVATE /Zi)
+
+   # Tell linker to include symbol data
+    set_target_properties(SPIRVDLL PROPERTIES
+        LINK_FLAGS "/INCREMENTAL:NO /DEBUG /OPT:REF /OPT:ICF"
+    )
+
+    # Set file name & location
+    set_target_properties(SPIRVDLL PROPERTIES
+        COMPILE_PDB_NAME SPIRVDLL
+        COMPILE_PDB_OUTPUT_DIR ${CMAKE_BINARY_DIR}
+    )
+endif()
+
 # --- end mock ---
 
 target_include_directories(SPIRVDLL
-- 
2.31.0

