From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nrudenko <nikita.rudenko@intel.com>
Date: Wed, 6 May 2020 13:38:34 +0300
Subject: [PATCH 3/8] Add -BSymbolic flag to libSPIRVDLL.so

---
 lib/SPIRV/CMakeLists.txt | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/SPIRV/CMakeLists.txt b/lib/SPIRV/CMakeLists.txt
index 00000000..00000000 100755
--- a/lib/SPIRV/CMakeLists.txt
+++ b/lib/SPIRV/CMakeLists.txt
@@ -49,6 +49,10 @@ llvm_update_compile_flags(SPIRVDLL)
 add_dependencies(SPIRVDLL intrinsics_gen LLVMAnalysis LLVMBitWriter LLVMCore LLVMSupport LLVMTransformUtils)
 target_link_libraries(SPIRVDLL LLVMAnalysis LLVMBitWriter LLVMCore LLVMSupport LLVMTransformUtils)
 
+if(NOT WIN32)
+  target_link_options(SPIRVDLL PRIVATE -Bsymbolic)
+endif()
+
 install(TARGETS SPIRVDLL
         EXPORT LLVMExports
         LIBRARY DESTINATION lib
-- 
2.31.0

